src/main.rs

```diff
@@ -658,99 +661,142 @@ fn choose_canonical_id(
         let writer = io::BufWriter::new(std::fs::File::create(path)?);
-        write_records(writer, records)?;
+        write_entries(writer, grouped_entries)?;
     } else {
         let stdout = io::stdout();
         if stdout.is_tty() {
             // do not write an extra newline if interactive
-            if !records.is_empty() {
-                write_records(stdout, records)?;
+            if !grouped_entries.is_empty() {
+                write_entries(stdout, grouped_entries)?;
             }
         } else {
             let writer = io::BufWriter::new(stdout);
-            write_records(writer, records)?;
+            write_entries(writer, grouped_entries)?;
         }
     };
 
     Ok(())
 }
 
 /// Iterate over records, writing the entries and warning about duplicates.
-fn write_records<W: io::Write, D: EntryData>(
+fn write_entries<W: io::Write, D: EntryData>(
     writer: W,
-    records: BTreeMap<RemoteId, NonEmpty<Entry<D>>>,
+    grouped_entries: BTreeMap<RemoteId, NonEmpty<Entry<D>>>,
 ) -> Result<(), serde_bibtex::Error> {
     let mut serializer = Serializer::unchecked(writer);
 
-    serializer.collect_seq(records.iter().flat_map(|(canonical, entries)| {
-        if entries.len() > 1 {
+    serializer.collect_seq(grouped_entries.iter().flat_map(|(canonical, entry_group)| {
+        if entry_group.len() > 1 {
             warn!(
                 "Multiple keys for '{canonical}': {}",
-                entries.iter().map(Entry::key).join(", ")
+                entry_group.iter().map(Entry::key).join(", ")
             );
         };
-        entries
+        entry_group
     }))
 }
 
-/// Validate and retrieve records.
-fn validate_and_retrieve<T: Iterator<Item = RecordId>>(
+/// Retrieve and validate BibTeX entries.
+fn retrieve_and_validate_entries<T: Iterator<Item = RecordId>>(
     citation_keys: T,
     record_db: &mut RecordDatabase,
     client: &HttpClient,
     ignore_null: bool,
 ) -> BTreeMap<RemoteId, NonEmpty<Entry<RawRecordData>>> {
-    let mut records: BTreeMap<RemoteId, NonEmpty<Entry<RawRecordData>>> = BTreeMap::new();
-
-    for (bibtex_entry, canonical) in citation_keys
-        .filter_map(|citation_key| {
-            match get_record(record_db, citation_key, client) {
-                Err(err) => {
-                    error!("{err}");
-                    None
-                }
-                Ok(GetRecordResponse::Exists(record)) => Some(record),
-                Ok(GetRecordResponse::NullRemoteId(remote_id)) => {
-                    if !ignore_null {
-                        error!("Null record: {remote_id}");
-                    }
-                    None
+    let valid_entries = citation_keys.filter_map(|citation_key| {
+        match get_record(record_db, citation_key, client) {
+            Err(err) => {
+                error!("{err}");
+                None
+            }
+            Ok(RecordResponse::Exists(record)) => get_bibtex_entry_from_record(record_db, record),
```

Comment:
Should this be wrapped inside a transaction? The call to the `linked keys` can be done inside the same transaction. Then the `get_valid_keys` can be implemented as a function on a `RecordRow` directly.

Reply:
There are several ways to merge the transactions. Which one do you have on your mind?

1. `RecordResponse::Exists` can carry a `Vec<String>` of alternative keys as well as the `Record`. But the alternatives are almost never needed.
2. `RecordResponse` can have another variant `InvalidKey(Vec<String>)`. This would mean moving the BibTeX validation logic into `record.rs`, which is pretty bad.
3. `RecordResponse::Exists` can forward the row transaction. This could have lifetime issues.

Reply:
I mean to use the `get_record_entry` function here, and match on the `RecordEntryResponse` (which passes the row back); and then either commit it or pass it down to the `get_bibtex_entry_from_record`.

Reply:
I see. So getting rid of `get_record` and `RecordResponse` completely then?

Reply:
Yeah exactly, I think this would work the best.

Reply:
And the `get_valid_keys` should have the same signature as, e.g., the `get_referencing_keys` function.

Reply:
Also in the `get_valid_keys` impl you can use `Vec::retain` to modify in-place (instead of filter / iter): https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain